PAPERLESS_ALLAUTH_TRUSTED_CLIENT_IP_HEADER effects in v3

### Description

Running Paperless in docker container on my Arch linux host macine. My reverse proxy is HAProxy on my firewall.
 
As per the docs, migration to v3 requires PAPERLESS_ALLAUTH_TRUSTED_CLIENT_IP_HEADER env being set otherwise I am met with "Server Error 500" when I attempt to login paperless through my reverse proxy using my public domain.

Without explicitly setting PAPERLESS_ALLAUTH_TRUSTED_CLIENT_IP_HEADER, here is what shows up in my logs when I attempt to login:
```
[2026-07-24 07:06:58,711] [ERROR] [django.request] Internal Server Error: /accounts/login/
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/asgiref/sync.py", line 555, in thread_handler
    raise exc_info[1]
  File "/usr/local/lib/python3.12/site-packages/django/core/handlers/exception.py", line 42, in inner
    response = await get_response(request)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/asgiref/sync.py", line 555, in thread_handler
    raise exc_info[1]
  File "/usr/local/lib/python3.12/site-packages/django/core/handlers/base.py", line 253, in _get_response_async
    response = await wrapped_callback(
               ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/asgiref/sync.py", line 504, in __call__
    ret = await asyncio.shield(exec_coro)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/asgiref/current_thread_executor.py", line 40, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/asgiref/sync.py", line 559, in thread_handler
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/django/views/generic/base.py", line 105, in view
    return self.dispatch(request, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/django/utils/decorators.py", line 48, in _wrapper
    return bound_method(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/allauth/decorators.py", line 14, in wrap
    resp = ratelimit.consume_or_429(request, action=action, **rl_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/allauth/core/ratelimit.py", line 72, in consume_or_429
    if not consume(request, *args, **kwargs):
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/allauth/core/ratelimit.py", line 41, in consume
    usage = _impl.consume(
            ^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/allauth/core/internal/ratelimit.py", line 172, in consume
    single_usage = _consume_single_rate(
                   ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/allauth/core/internal/ratelimit.py", line 133, in _consume_single_rate
    cache_key = get_cache_key(request, action=action, rate=rate, key=key, user=user)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/allauth/core/internal/ratelimit.py", line 101, in get_cache_key
    source = ("ip", get_adapter().get_client_ip(request))
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/allauth/account/adapter.py", line 817, in get_client_ip
    ip = get_client_ip(request)
         ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/allauth/core/internal/httpkit.py", line 217, in get_client_ip
    ip = get_client_ip_from_xff(request)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/allauth/core/internal/httpkit.py", line 203, in get_client_ip_from_xff
    raise ImproperlyConfigured(
django.core.exceptions.ImproperlyConfigured: ALLAUTH_TRUSTED_PROXY_COUNT does not match X-Forwarded-For
```
Note the login page itself loads. This only happens after attempting to submit credentials at the login page (regardless if they are valid or not). It is throws a server error 500 immediately and won't even validate the login.

Setting "PAPERLESS_ALLAUTH_TRUSTED_CLIENT_IP_HEADER=X-Forwarded-For" solves the problem and I can login and use paperless through the reverse proxy again, as the documentation notes. (Note: Setting ALLAUTH_TRUSTED_PROXY_COUNT itself does not do anything to solve this for me)

However if "PAPERLESS_ALLAUTH_TRUSTED_CLIENT_IP_HEADER=X-Forwarded-For" is set, logging in locally fails leading to Forbidden 403 error. The following shows up in the logs when trying to login into paperless locally (ie http://192.168.1.1:8000):
```
[2026-07-24 07:23:53,508] [WARNING] [django.request] Forbidden (Permission denied): /accounts/login/
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/asgiref/sync.py", line 555, in thread_handler
    raise exc_info[1]
  File "/usr/local/lib/python3.12/site-packages/django/core/handlers/exception.py", line 42, in inner
    response = await get_response(request)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/asgiref/sync.py", line 555, in thread_handler
    raise exc_info[1]
  File "/usr/local/lib/python3.12/site-packages/django/core/handlers/base.py", line 253, in _get_response_async
    response = await wrapped_callback(
               ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/asgiref/sync.py", line 504, in __call__
    ret = await asyncio.shield(exec_coro)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/asgiref/current_thread_executor.py", line 40, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/asgiref/sync.py", line 559, in thread_handler
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/django/views/generic/base.py", line 105, in view
    return self.dispatch(request, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/django/utils/decorators.py", line 48, in _wrapper
    return bound_method(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/allauth/decorators.py", line 14, in wrap
    resp = ratelimit.consume_or_429(request, action=action, **rl_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/allauth/core/ratelimit.py", line 72, in consume_or_429
    if not consume(request, *args, **kwargs):
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/allauth/core/ratelimit.py", line 41, in consume
    usage = _impl.consume(
            ^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/allauth/core/internal/ratelimit.py", line 172, in consume
    single_usage = _consume_single_rate(
                   ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/allauth/core/internal/ratelimit.py", line 133, in _consume_single_rate
    cache_key = get_cache_key(request, action=action, rate=rate, key=key, user=user)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/allauth/core/internal/ratelimit.py", line 101, in get_cache_key
    source = ("ip", get_adapter().get_client_ip(request))
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/allauth/account/adapter.py", line 819, in get_client_ip
    raise PermissionDenied("Unable to determine client IP address")
django.core.exceptions.PermissionDenied: Unable to determine client IP address
```
Clearly it won't allow local login anymore because it is looking for X-Forwarded-For header for the client IP which will not happen using local access.

Note I have set both my public domain and host machine IP in the ALLOWED_HOSTS env so that both methods would work in v2. This results in two issues for me personally. Nextcloud on the host machine will not longer be able login new clients locally without going through the reverse proxy (undesirable since my RA is on another machine). This is also preventing HAProxy from performing my preferred health check config on paperless. It checks the headers on /accounts/login/ for status 200 before HAProxy will allow paperless to be reached publicly. Sure I could change it to GET but that shouldn't necessary for health checking. When PAPERLESS_ALLAUTH_TRUSTED_CLIENT_IP_HEADER is set, it will receive 403 instead of 200.

I also believe there maybe security issue in this as well. Because if the client is already logged, therefore bypassing the login page, the client is able to use the container even if PAPERLESS_ALLAUTH_TRUSTED_CLIENT_IP_HEADER is set or unset and the container is reloaded.

Meaning if it PAPERLESS_ALLAUTH_TRUSTED_CLIENT_IP_HEADER is unset, I cannot login via the reverse proxy but I can login locally. Then I set the env and reload the container. The locally logged in client remains logged in and can continue using paperless without encountering the login issue or errors. This is true vise-versa. If the PAPERLESS_ALLAUTH_TRUSTED_CLIENT_IP_HEADER is set, logging in through the RA works but locally otherwise fails. If I log in a client through the RA then unset the env and reload the container, the clients that already logged in through the RA previously can continue to use the container. The whole client IP issue only seems to occur when submitting credentials at the login page. What exactly is the intended purpose of PAPERLESS_ALLAUTH_TRUSTED_CLIENT_IP_HEADER?

### Steps to reproduce

Configure paperless to use HAproxy with public domain and also set the local machine IP in the ALLOWED_HOSTS for local access.

Set "PAPERLESS_ALLAUTH_TRUSTED_CLIENT_IP_HEADER=X-Forwarded-For" to allow logging in through the RA public domain and logging in via local IP will fail.

Unset "PAPERLESS_ALLAUTH_TRUSTED_CLIENT_IP_HEADER=X-Forwarded-For" to allow logging in via local IP and logging in via public domain will fail.

Once a client is logged it, it stays logged in even if the env is set or unset and the container is reload. I think this has security implications if the client is not meant to be accessing anymore based on the env setting?

### Webserver logs

```bash
Logs above.
```

### Browser logs

```bash

```

### Services logs

```bash

```

### Paperless-ngx version

3.0.2

### Host OS

Archlinux

### Installation method

Docker - official image

### System status

```json

```

### Browser

Firefox

### Configuration changes

_No response_

### Please confirm the following

- [x] I believe this issue is a bug that affects all users of Paperless-ngx, not something specific to my installation.
- [x] This issue is not about the OCR or archive creation of a specific file(s). Otherwise, please see above regarding OCR tools.
- [x] I have already searched for relevant existing issues and discussions before opening this report.
- [x] I have updated the title field above with a concise description.

--- a/docs/configuration.md
+++ b/docs/configuration.md
@@ -522,22 +522,36 @@ do CORS calls. Set this to your public domain name.
 fail2ban with log entries for failed authorization attempts. Value should be
 IP address(es).
 
-    This setting also controls allauth's
-    [`ALLAUTH_TRUSTED_PROXY_COUNT`](https://docs.allauth.org/en/latest/account/configuration.html),
-    which is set to the number of proxies listed here. Without this,
-    allauth cannot determine the client IP address for rate limiting when
-    running behind a reverse proxy, resulting in a `403 Forbidden` on login.
+    By default, this setting also controls allauth's trusted proxy count,
+    which is set to the number of proxies listed here. Override that default
+    with [`PAPERLESS_ALLAUTH_TRUSTED_PROXY_COUNT`](#PAPERLESS_ALLAUTH_TRUSTED_PROXY_COUNT)
+    when the list length does not match the number of proxy hops.
 
     Defaults to empty string.
+
+#### [`PAPERLESS_ALLAUTH_TRUSTED_PROXY_COUNT=<integer>`](#PAPERLESS_ALLAUTH_TRUSTED_PROXY_COUNT) {#PAPERLESS_ALLAUTH_TRUSTED_PROXY_COUNT}
+
+: Sets allauth's
+[`ALLAUTH_TRUSTED_PROXY_COUNT`](https://docs.allauth.org/en/latest/common/rate_limits.html#configuration).
+This is the number of trusted proxy **hops** represented in each
+`X-Forwarded-For` header, not the number of IP addresses through which those
+proxies may be reached. For example, a single dual-stack proxy is one hop even
+when its IPv4 and IPv6 addresses are both listed in
+[`PAPERLESS_TRUSTED_PROXIES`](#PAPERLESS_TRUSTED_PROXIES).
+
+    Only trust `X-Forwarded-For` when untrusted clients cannot connect directly
+    to Paperless-ngx.
+
+    Defaults to the number of entries in `PAPERLESS_TRUSTED_PROXIES`.
 
 #### [`PAPERLESS_ALLAUTH_TRUSTED_CLIENT_IP_HEADER=<header-name>`](#PAPERLESS_ALLAUTH_TRUSTED_CLIENT_IP_HEADER) {#PAPERLESS_ALLAUTH_TRUSTED_CLIENT_IP_HEADER}
 
 : Sets allauth's
-[`ALLAUTH_TRUSTED_CLIENT_IP_HEADER`](https://docs.allauth.org/en/latest/account/configuration.html).
+[`ALLAUTH_TRUSTED_CLIENT_IP_HEADER`](https://docs.allauth.org/en/latest/common/rate_limits.html#configuration).
 Use this when your reverse proxy sets a dedicated header for the real
 client IP instead of `X-Forwarded-For`, for example `X-Real-IP` (nginx)
 or `CF-Connecting-IP` (Cloudflare). When set, this takes precedence over
-[`PAPERLESS_TRUSTED_PROXIES`](#PAPERLESS_TRUSTED_PROXIES).
+`PAPERLESS_ALLAUTH_TRUSTED_PROXY_COUNT`.
 
     Defaults to none.
 

--- a/docs/migration-v3.md
+++ b/docs/migration-v3.md
@@ -324,8 +324,11 @@ option since v1.8.0.
 Allauth changed how it determines the client IP address for login rate limiting. Users running
 behind a reverse proxy may need to set
 [`PAPERLESS_TRUSTED_PROXIES`](configuration.md#PAPERLESS_TRUSTED_PROXIES),
+[`PAPERLESS_ALLAUTH_TRUSTED_PROXY_COUNT`](configuration.md#PAPERLESS_ALLAUTH_TRUSTED_PROXY_COUNT),
 [`PAPERLESS_ALLAUTH_TRUSTED_CLIENT_IP_HEADER`](configuration.md#PAPERLESS_ALLAUTH_TRUSTED_CLIENT_IP_HEADER),
-or both, to avoid `403 Forbidden` errors on login.
+or a combination of these settings to avoid `403 Forbidden` errors on login.
+The proxy count is the number of proxy hops in `X-Forwarded-For`, which may
+differ from the number of configured proxy IP addresses.
 
 ## Minimum CPU Requirements (NumPy Baseline)
 

--- a/src/paperless/settings/__init__.py
+++ b/src/paperless/settings/__init__.py
@@ -460,11 +460,22 @@ def _parse_paperless_url():
 
 PAPERLESS_URL = _parse_paperless_url()
 
+
+def _get_allauth_trusted_proxy_count(trusted_proxies: list[str]) -> int:
+    count = get_int_from_env(
+        "PAPERLESS_ALLAUTH_TRUSTED_PROXY_COUNT",
+        len(trusted_proxies),
+    )
+    if count < 0:
+        raise ImproperlyConfigured(
+            "PAPERLESS_ALLAUTH_TRUSTED_PROXY_COUNT must be zero or greater",
+        )
+    return count
+
+
 # For use with trusted proxies
 TRUSTED_PROXIES = get_list_from_env("PAPERLESS_TRUSTED_PROXIES")
-# Derive allauth's proxy count from the same list so X-Forwarded-For is trusted
-# correctly when users have configured PAPERLESS_TRUSTED_PROXIES.
-ALLAUTH_TRUSTED_PROXY_COUNT = len(TRUSTED_PROXIES)
+ALLAUTH_TRUSTED_PROXY_COUNT = _get_allauth_trusted_proxy_count(TRUSTED_PROXIES)
 ALLAUTH_TRUSTED_CLIENT_IP_HEADER = os.getenv(
     "PAPERLESS_ALLAUTH_TRUSTED_CLIENT_IP_HEADER",
 )

--- a/src/paperless/tests/settings/test_settings.py
+++ b/src/paperless/tests/settings/test_settings.py
@@ -3,7 +3,9 @@ from unittest import TestCase
 from unittest import mock
 
 import pytest
+from django.core.exceptions import ImproperlyConfigured
 
+from paperless.settings import _get_allauth_trusted_proxy_count
 from paperless.settings import _get_search_language_setting
 from paperless.settings import _parse_paperless_url
 from paperless.settings import default_threads_per_worker
@@ -33,6 +35,31 @@ class TestThreadCalculation(TestCase):
                 self.assertGreaterEqual(default_threads, 1)
 
                 self.assertLessEqual(default_workers * default_threads, i)
+
+
+def test_allauth_trusted_proxy_count_defaults_to_trusted_proxies(
+    monkeypatch: pytest.MonkeyPatch,
+) -> None:
+    monkeypatch.delenv("PAPERLESS_ALLAUTH_TRUSTED_PROXY_COUNT", raising=False)
+
+    assert _get_allauth_trusted_proxy_count(["proxy-v4", "proxy-v6"]) == 2
+
+
+def test_allauth_trusted_proxy_count_can_be_overridden(
+    monkeypatch: pytest.MonkeyPatch,
+) -> None:
+    monkeypatch.setenv("PAPERLESS_ALLAUTH_TRUSTED_PROXY_COUNT", "1")
+
+    assert _get_allauth_trusted_proxy_count(["proxy-v4", "proxy-v6"]) == 1
+
+
+def test_allauth_trusted_proxy_count_rejects_negative_values(
+    monkeypatch: pytest.MonkeyPatch,
+) -> None:
+    monkeypatch.setenv("PAPERLESS_ALLAUTH_TRUSTED_PROXY_COUNT", "-1")
+
+    with pytest.raises(ImproperlyConfigured, match="must be zero or greater"):
+        _get_allauth_trusted_proxy_count([])
 
 
 @pytest.mark.parametrize(

